cpp/src/arrow/c/abi.h

```diff
@@ -319,6 +319,11 @@ struct ArrowAsyncProducer {
   // on_error callback on the async stream handler.
   void (*cancel)(struct ArrowAsyncProducer* self);
 
+  // Any additional metadata tied to a specific stream of data. This must either be NULL
+  // or a valid pointer to metadata which is encoded in the same way schema metadata
+  // would be. Non-null metadata must be valid for the lifetime of this object.
```

Comment:
nit: can we keep the context about what kind of metadata might be useful to pass, and the encoding of the string?

Reply:
Also we should note it may only be valid after the schema callback?

Reply:
This entire struct is only valid after the schema callback, so noting that again for just this field seems redundant. What kind of context are you thinking of? An example like "a producer might use this to communicate the total number of rows and/or batches in the stream if that is known? "

Reply:
Basically the context we had earlier, yes

Reply:
This entire struct must be valid *before* the `on_schema` call, not after.

I'll add the comment about the context etc.